iceoryx_hoofs/source/posix_wrapper/periodic_timer.cpp

```diff
@@ -0,0 +1,188 @@
+{
+    struct timespec ts;
+    auto result =
+        posixCall(clock_gettime)(CLOCK_REALTIME, &ts).failureReturnValue(-1).evaluate();
+
+    if (result.has_error())
+    {
+        return createErrorCodeFromErrNo(result.get_error().errnum);
+    }
+
+    return cxx::success<units::Duration>(ts);
+}
+
+cxx::expected<iox::posix::WaitResult, TimerErrorCause> PeriodicTimer::wait(TimerCatchupPolicy policy) noexcept
+{
+    // To check if the TIMER is active (if the sempahore is acquired)
+    auto semValue = m_waitSemaphore.getValue();
+    cxx::Ensures(!semValue.has_error());
+    if (semValue.value() == iox::posix::SEM_ACQUIRED)
+    {
+        auto currentTime = now();
+        cxx::Ensures(!currentTime.has_error());
+        if (currentTime.value() > m_timeForNextActivation)
+        {
+            switch (policy)
+            {
+            case iox::posix::TimerCatchupPolicy::IMMEDIATE_TICK:
+            {
+                m_timeForNextActivation = currentTime.value();
+                m_waitResult.state = iox::posix::TimerState::TICK;
+                return cxx::success<iox::posix::WaitResult>(m_waitResult);
+            }
+
+            case iox::posix::TimerCatchupPolicy::SKIP_TO_NEXT_TICK:
+            {
+                auto delay = currentTime.value() - m_timeForNextActivation; // calculate the time delay
+                if (delay > m_interval)
+                {
+                    auto totalSlotToBeSkipped =
+                        delay.toMilliseconds() / m_interval.toMilliseconds(); // calculate the total slots missed
+                    m_timeForNextActivation =
+                        m_timeForNextActivation + m_interval * totalSlotToBeSkipped; // Skip to the next activation
+                }
+                else
+                {
+                    m_timeForNextActivation = m_timeForNextActivation + m_interval; // Skip to the next activation
+                }
+                auto timeDiff = m_timeForNextActivation - currentTime.value(); // calculate remaining time for activation
+                auto waitResult = m_waitSemaphore.timedWait(timeDiff);
+                if (waitResult.has_error())
+                {
+                    return cxx::error<TimerErrorCause>(TimerErrorCause::INTERNAL_LOGIC_ERROR);
+                }
+                else
+                {
+                    m_waitResult.state = iox::posix::TimerState::TICK;
+                    return cxx::success<iox::posix::WaitResult>(m_waitResult);
+                }
+            }
+
+            case iox::posix::TimerCatchupPolicy::HOLD_ON_DELAY:
+            {
+                auto timeDiff = currentTime.value() - m_timeForNextActivation; // Calculate the time delay
+                m_waitResult.state = iox::posix::TimerState::DELAY;
+                m_waitResult.timeDelay = timeDiff;
+                return cxx::success<iox::posix::WaitResult>(m_waitResult);
+            }
```

Comment:
I think this should not return in the happy path, like explained in https://github.com/eclipse-iceoryx/iceoryx/pull/754#discussion_r622086949

@mossmaurice @elfenpiff what do you think?